### PR TITLE
Remove precision and sign loss warnings from util.img_as_

### LINKED ISSUE #2602: Add warn keyword to img_as_* functions
Image conversion from float to uint8 is very, very common, and usually people know what they're doing when using it, and don't appreciate the annoying warning. A brief discussion in https://github.com/scikit-image/scikit-image/issues/543#issuecomment-292221966 suggests that users might like a `img_as_uint8(..., warn=False)` keyword argument as a way to suppress the warnings.

The same would be true for `imsave` and low contrast images.

Another idea would be to only warn when the apparent dynamic range is dramatically compressed, even when `warn=True`.

Discuss. =P

### PR DESCRIPTION
**Pull request by jni**
## Description
See discussions:
#543 (comment)
#2602

Specifically, @stefanv said [1]_:

> Why don't we just remove this warning entirely? It seemed like the
> proper thing to do at the time, but it's never been useful.

Therefore, I set the default to False, but leave the option for warning
there since it's potentially useful to some. When and if we stop doing
automatic dtype conversions (see #3009), we might be able to remove
warnings altogether.

..[1]: #2602 (comment) 

The PR touches a lot of lines so I suggest looking at individual commits when reviewing. Thanks @scikit-image/core!

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests

## References
Closes #2602 

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
There don't seem to be any tests that actually test that the warnings are correct. Maybe you can adapt these: https://github.com/scikit-image/scikit-image/pull/3055/files#diff-74fec0b61086721cf4ec492e880e1d6cR153

**Comment by jni** [comment]
@hmaarrfk I just left a few tests open where the warnings are on. Search for `expected_warnings.*precision`, or `warn_on_precision_loss=True`.

File: skimage/filters/rank/generic.py [hunk 1]
```diff
@@ -51,7 +51,7 @@
 import functools
 import numpy as np
 from scipy import ndimage as ndi
-from ... import img_as_ubyte
+from ...util import img_as_ubyte
```
[the remark this answers is not in the document]
Comment (jni) [reply skimage/filters/rank/generic.py]: This is very much a fair point. I'll let @scikit-image/core comment before I make that change, but I am :+1: on this suggestion.

**Comment by hmaarrfk** [comment]
@jni, they just don't seem systematically tested.

See this issue for example
https://github.com/scikit-image/scikit-image/issues/3080

https://github.com/scikit-image/scikit-image/pull/3055 had fixed it, but there was too much magic going on in one of the conversions, so I stopped pushing for it.

**Comment by jni** [comment]
:scream: wow. Yeah, I don't particularly want to get deep into the logic of `convert` for this PR...

**Comment by alexandrejaguar** [comment]
Hi all,
I think we should believe in the users a little bit here. Wouldn't it be easier for us to only remove the warning, as @stefanv suggested, and pointing the precautions users should have when dealing with data conversion in a tutorial, or something like that?
Personally, I don't see a situation where maintaining the hidden warning would be useful. OK, maybe one or two cases popped on my mind now, but I still think we should let that in the users hands; my 2 ¢.

File: doc/source/user_guide/data_types.rst [hunk 2]
```diff
@@ -176,7 +181,7 @@ a custom function that requires a particular dtype, you should call one of the
 dtype conversion functions (here, ``func1`` and ``func2`` are ``skimage``
 functions)::
 
-   >>> from skimage import img_as_float
+   >>> from skimage.util import img_as_float
```
Comment (stefanv) [inline doc/source/user_guide/data_types.rst]: Why these changes?  I think those function remain exposed on the default namespace, without plans to change that any time soon or ever.

Comment (jni) [reply doc/source/user_guide/data_types.rst]: @stefanv I object to the second clause in that sentence. =P 

I want this to be deprecated in the same way that `np.matrix` was "deprecated" for a long time before any messages appeared. ie, discouraged. For me it makes much more sense for these functions to be in `util` than "polluting" the top-level domain, which is otherwise only modules and the `data_dir` variable.

**Comment by jni** [comment]
Ok, to quote the great Samuel L Jackson, hold on to your butts!

Since we have two core devs in favour of removing altogether, and no one (as far as I can tell) strongly against it, I've indeed removed all the warnings wholesale. Let's see if anyone complains... ;)

**Comment by sciunto** [comment]
Travis complains :D

File: doc/source/user_guide/data_types.rst (hunk 2 above)
Comment (stefanv) [reply doc/source/user_guide/data_types.rst]: I should have been clearer, I guess: without your plans having been made known to the rest of us? :)

Seriously, though, those functions are strewn throughout *all* our lecture materials, so it would be quite painful to get rid of it.  Now, w.r.t. discouragement, I could go for that, but do you have a specific rationale for why `skimage.util.img_as_float` is better than `skimage.img_as_float`, other than the slight aesthetic benefit of having only submodules in the namespace?  I guess then we'll also need to get rid of `skimage.test()`, which leaves only submodules + `__version__`.

Comment (hmaarrfk) [reply doc/source/user_guide/data_types.rst]: I once listened to a talk that was advocate against this kind of nested API. They were saying that we should strive for a "flat" API design. not sure how applicable that is to skimage.

a flat API can be easier to learn, at least easier to remember what to import from.

Comment (jni) [reply doc/source/user_guide/data_types.rst]: I was under the impression that this idea had been lightly discussed elsewhere, but I can't find any evidence of that, so, sorry about that.

I don't consider "this is the way we've always done it" to be a big deal, especially in the context of a soft deprecation. There is zero urgency in this removal, it is just something that I think we should do "right" going forward, and especially in our own code base. There's outdated tutorials for everything online, and this should not hold us back, any more than "there's lots of tutorials and books written in Python 2" should hold back the transition to 3. Now, I was auditing the use of these functions in our codebase so I took the opportunity to update these. (And note that we had both conventions floating about in our codebase, which is a jarring inconsistency.)

I actually think `test` and `lookfor` and maybe even `data_dir` should stay in the top level as being "package-wide" things. Is `img_as_float` something that refers to our package as a whole, or a utility function? For me it's a slam dunk for the latter. Put another way: does `img_as_float` belong in the same bucket as `lookfor` and `test`?

**Comment by jni** [comment]
@sciunto wow it's funny how a refactor can uncover unexpected badness! There were a lot of `from skimage._shared.testing import expected_warnings`, but that function is in `_shared._warnings` — it was in `_shared.testing` only as a side effect of it being used in that file. When I removed that use, all those imports broke. :man_facepalming: 

Anyway, :crossed_fingers: !

File: doc/source/user_guide/data_types.rst (hunk 2 above)
Comment (stefanv) [reply doc/source/user_guide/data_types.rst]: Perhaps what bothers me is the nondescript `util` name, which will not help users discover these.  They could potentially live in `io`, or in their own submodule.

You are arguing that these functions are not-package wide, which I agree with.  But another perspective could be that they belong there because they are invoked on every single pipeline, and should therefore be highly visible.

Hopefully they will be needed less as we embrace float images without defined ranges, wherever possible.

Comment (jni) [reply doc/source/user_guide/data_types.rst]: Hey I'm the one who argued for a flat namespace to help discoverability and was vetoed. =P

More seriously, as you say, I hope that long term they are not so indispensable.

Finally, perhaps a new module, `skimage.dtypes`? Or somesuch?

**Comment by jni** [comment]
Oh man, Travis is **so** unhappy...

File: doc/source/user_guide/data_types.rst (hunk 2 above)
Comment (hmaarrfk) [reply doc/source/user_guide/data_types.rst]: It seems that maybe you should restrict the scope of the PR. I don't think there is much resistance to any of the proposed changes, but it seems that together, they are hard to justify at once as some are cosmetic, and some are behavioural.

**Comment by soupault** [comment]
@jni Could you rebase, please?

**Comment by jni** [comment]
I doubt it'll help but sure. =D Hold please...

**Comment by jni** [comment]
Erm, where did Travis go?

**Comment by jni** [comment]
@jakirkham thanks for the `--allow-empty` tip! ;) :point_up: 

Hmmm, your instincts were right, @soupault, only two failures!

https://travis-ci.org/scikit-image/scikit-image/jobs/490881001#L2027-L2148

I don't understand either of them though... :joy:

**Comment by jni** [comment]
@scikit-image/core I think my latest two commits should fix the remaining two errors. Please give it another round of review!

**Comment by jni** [comment]
@hmaarrfk eh, I'm in favour of leaving that to #3009. ;)

File: skimage/filters/rank/generic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply skimage/filters/rank/generic.py]: Do you mind expanding on the issue here? What is being proposed?

[the remark this answers is not in the document]
Comment (jni) [reply skimage/filters/rank/generic.py]: @stefanv I think @hmaarrfk is making a distinction between precision loss warnings stemming from IO and user calls to `img_as_ubyte`, which should be silenced, and precision loss warnings because rank filters convert your images to uint8, and now do so completely silently.

Having said that I would like to punt on this issue, as I don't consider it that critical or urgent. (See main conversation thread on the PR.)

[the remark this answers is not in the document]
Comment (soupault) [reply skimage/filters/rank/generic.py]: >convert your images to uint8, and now do so completely silently

I think that happens in multiple other places also. 👍 to issue the warning.

[the remark this answers is not in the document]
Comment (jni) [reply skimage/filters/rank/generic.py]: Done.

**Comment by jni** [comment]
@scikit-image/core the napari team just ran into the "excessive warnings" issue. This was approved by @hmaarrfk already. Could someone else review and merge? I've rebased on master, the tests all pass locally, let's see what Travis & Co say, and hopefully get merging!

**Comment by soupault** [comment]
Thanks @jni ! Nice piece of work!

**Comment by jni** [comment]
Thanks @soupault! Nice work with the title and final touches! =)

**Comment by hmaarrfk** [comment]
It is almost worth release 0.16 just for this.

**Comment by jni** [comment]
:joy: I agree. Actually I'd like to get it out soon after SciPy, but I probably won't be able to get to it until I'm back in Melbourne.
